stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
 
-/**************************************************************************************************/
 
 class notification_queue {
     using lock_t = std::unique_lock<std::mutex>;
     std::deque<task<void()>> _q;
-    bool _done{false};
-    std::mutex _mutex;
-    std::condition_variable _ready;
+    std::mutex _queue_mutex;
+    std::condition_variable* _ready{nullptr};
 
 public:
+    void set_context(std::condition_variable& cv) {
+        _ready = &cv;
+    }
+
     bool try_pop(task<void()>& x) {
-        lock_t lock{_mutex, std::try_to_lock};
+        lock_t lock{_queue_mutex, std::try_to_lock};
         if (!lock || _q.empty()) return false;
         x = std::move(_q.front());
         _q.pop_front();
         return true;
     }
 
+    bool pop(task<void()>& x) {
+        lock_t lock{_queue_mutex};
+        if (_q.empty()) return false;
+        x = std::move(_q.front());
+        _q.pop_front();
+        return true;
+    }
+
+
     template <typename F>
     bool try_push(F&& f) {
         {
-            lock_t lock{_mutex, std::try_to_lock};
+            lock_t lock{_queue_mutex, std::try_to_lock};
             if (!lock) return false;
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
-        return true;
-    }
-
-    void done() {
-        {
-            lock_t lock{_mutex};
-            _done = true;
-        }
-        _ready.notify_all();
-    }
-
-    bool pop(task<void()>& x) {
-        lock_t lock{_mutex};
-        while (_q.empty() && !_done)
-            _ready.wait(lock);
-        if (_q.empty()) return false;
-        x = std::move(_q.front());
-        _q.pop_front();
+        _ready->notify_all();
         return true;
     }
 
     template <typename F>
     void push(F&& f) {
         {
-            lock_t lock{_mutex};
+            lock_t lock{_queue_mutex};
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
+        _ready->notify_all();
     }
 };
 
 /**************************************************************************************************/
 
-class task_system {
+class priority_task_system {
+    using lock_t = std::unique_lock<std::mutex>;
+
     const unsigned _count{std::thread::hardware_concurrency()};
-    std::vector<std::thread> _threads;
-    std::vector<notification_queue> _q{_count};
+
+    struct thread_context
+    {
+        std::mutex mutex;
+        std::condition_variable ready;
+        std::thread thread;
+    };
+    std::vector<thread_context> _thread_contexts{_count};
+    std::array<std::vector<notification_queue>, 3> _q;
     std::atomic<unsigned> _index{0};
+    std::atomic_bool _done{false};
 
     void run(unsigned i) {
         while (true) {
+            begin:
             task<void()> f;
 
-            for (unsigned n = 0; n != _count * 32; ++n) {
-                if (_q[(i + n) % _count].try_pop(f)) break;
+            for (auto& q : _q) {
```

Comment:
@FelixPetriconi Perhaps rather than using labels we could add `bool do_try_pop(task<void()>& x)` and `bool do_pop(task<void()>& x)` then  a simple `continue` if they return true?

Reply:
I wanted to avoid here extra variables, because this run() function is time critical.
I know Dijkstra "Go To Statement Considered Harmful" statement. But there are places where this is a valuable tool. I think, this here is one.

Reply:
If that is the case then maybe we should reuse the `if`s inside the `notification_queue` and use a chain or responsibility to avoid extra comparisons. So the `task_system` would have a `while` loop that tries to get/run a task. First we `try_pop` and if is succeeds it executes `f` in place. If it fails then it calls `pop`. If pop succeeds it executes `f` in place if it fails then it calls `wait`.

Reply:
If it is time critical, couldn't we add the routines @aaronalbers suggests an inline them? I would hope this routine profile is far smaller than the time taken in `f()` anyhow...